### ISSUE TITLE
Status of MotoSim?

### ISSUE THREAD

**Issue by gavanderhoorn** [ra-mtp-ntnu/moto#15]
There seems to be support for a form of 'simulation', which starts a set of server threads which implement subsets of the Simple Message services and topics supported by MotoROS.

It doesn't appear to be complete though.

I'm wondering what the status of it is, and whether there are any plans to complete it?

**Comment by tingelst** [comment]
Sorry for the late response @gavanderhoorn, Ive been  out of office on parental leave for the last months.

Completion of the motoros simulator is on my agenda. Are there any specific features that you are looking for?

**Comment by gavanderhoorn** [comment]
I've mainly been using this as a substitute for my controller.

The behaviour of the scripts in the repository is / wasn't close enough to how MotoROS behaves, so I had to make some changes.

It's been a while though, so I'll take a look at what I changed and potentially contribute those changes.

**Comment by tingelst** [comment]
> The behaviour of the scripts in the repository is / wasn't close enough to how MotoROS behaves, so I had to make some changes.

Yes the behavior is not as the real driver. I wrote it mainly to test the motion server during the initial development when I didnt have access to the physical robot.

> It's been a while though, so I'll take a look at what I changed and potentially contribute those changes.

That would be great!

**Comment by rr-tom-noble** [comment]
Hey @tingelst,

Following on from [this](https://github.com/ros-industrial/motoman/discussions/510) discussion, @gavanderhoorn pointed me in the direction of this repo. The project I'm working on makes heavy use of point streaming (which, admittedly, comes from a fork of the motoman repo, but I believe still just uses MotoROS under the hood). Finding issues when we're testing on real hardware, late in the development process, is a major source of headaches. A fully functional mock MotoROS server sounds like it'd be awesome for catching those bugs much earlier.

I'd be happy to contribute, although my knowledge of MotoROS is a little weak, so any info / material to get up-to-speed would be much appreciated. I'll have a look over the source as well :slightly_smiling_face:

**Comment by tingelst** [comment]
Hey @rr-tom-noble!

I think your way of testing the original ROS driver towards the MotoSim implementation here is a great start for further development. 

Could you please elaborate on the kinds of problems you would like to reduce by testing towards a simulated interface? This might help to narrow the scope of the development efforts.

**Comment by riv-tnoble** [comment]
The most common type of issue we've seen is where a motion executes fine in simulation, but motoman raises an error when executing on the real controller.

Again, we're using a forked point streaming branch, so maybe that's uncommon for trajectories that are not streamed. Off the top of my head I can think of a few common errors:
 - Current joints are too far from start of trajectory
 - Time from start must be greater than previous point
 - General errors from the controller

Being able to see those types of errors in simulation would allow us to pick up on them earlier

**Comment by gavanderhoorn** [comment]
I've added most of those to my local copy, just haven't had time to upstream changes.

Also added multi-group support.

> General errors from the controller

this is too vague, and will also take too much time.
